bfabric/bfabric_config.py

```diff
@@ -27,73 +30,146 @@ class BfabricConfig:
-        base_url=values.get("_WEBBASE"),
-        application_ids=values.get("_APPLICATION"),
-        job_notification_emails=values.get("_JOB_NOTIFICATION_EMAILS"),
-    )
-    config = BfabricConfig(**{k: v for k, v in args.items() if v is not None})
-    if "_LOGIN" in values and "_PASSWD" in values:
-        auth = BfabricAuth(login=values["_LOGIN"], password=values["_PASSWD"])
+            logger.log(20, "found BFABRICPY_CONFIG_ENV = " + config_env)
     else:
+        logger.log(20, "config environment specified explicitly as " + config_env)
+
+    if config_env not in config_dict:
+        raise IOError("The requested config environment", config_env, "is not present in the config file")
+
+    return config_env, config_dict[config_env]
+
+def _have_all_keys(d: dict, l: list) -> bool:
+    """True if all elements in list l are present as keys in dict d, otherwise false"""
+    return all([k in d for k in l])
+
+def _parse_dict(d: dict, mandatory_keys: list, optional_keys: list = None, error_prefix: str = None):
+    """
+    Returns a copy of an existing dictionary, only keeping mandatory and optional keys
+    If a mandatory key is not found, an exception is raised
+    :param d:                 Starting dictionary
+    :param mandatory_keys:    A list of mandatory keys
+    :param optional_keys:     A list of optional keys
+    :param error_prefix:      A string to print if a mandatory key is not found
+    :return:                  Copy of a starting dictionary, only containing mandatory and optional keys
+    """
+    d_rez = {}
+
+    # Get all mandatory fields, and complain if not found
+    for k in mandatory_keys:
+        if k in d:
+            d_rez[k] = d[k]
+        else:
+            raise ValueError(error_prefix + k)
+
+    # Get all optional fields
+    if optional_keys is not None:
+        for k in optional_keys:
+            if k in d:
+                d_rez[k] = d[k]
+
+    # Ignore all other fields
+    return d_rez
+
+def read_bfabricrc_py(config_path: str, config_env: str = None,
+                      optional_auth: bool = False) -> Tuple[BfabricConfig, Optional[BfabricAuth]]:
+    """
+    Reads bfabricpy.yml file, parses it, extracting authentication and configuration data
+    :param config_path:   Path to the configuration file. It is assumed the file exists
+    :param config_env:    Configuration environment to use. If not given, it is deduced.
+    :param optional_auth: Whether authentication is optional.
+        If not, both login and password must be present in the config file, otherwise an exception is thrown
+        If yes, missing login and password would result in authentication class being None, but no exception
+    :return: Configuration and Authentication class instances
+    """
+
+
+    config_env_final, config_dict = _read_config_env_as_dict(config_path, config_env=config_env)
+
+    error_prefix = "Config environment " + config_env_final + " does not have a compulsory field: "
```

Comment:
Btw, I'm commenting theses f-string opportunities as I see them, but at some point for existing code we can use a tool to replace all relevant instances.

```suggestion
    error_prefix = f"Config environment {config_env_final} does not have a compulsory field: "
```

Reply:
I never learned f-strings when I was learning python, and now I'm in a habbit of using string concatenation the noob way. Sorry. I'll try to improve